```
2021-01-28T12:40:15.6058709Z === RUN   TestNameExpand
2021-01-28T12:40:15.6062905Z     name_test.go:44: test 4: result: expected: [s3:ObjectCreated:CompleteMultipartUpload s3:ObjectCreated:Copy s3:ObjectCreated:Post s3:ObjectCreated:Put s3:ObjectCreated:PutRetention s3:ObjectCreated:PutLegalHold s3:Replication:OperationCompletedReplication s3:Replication:OperationFailedReplication], got: [s3:ObjectCreated:CompleteMultipartUpload s3:ObjectCreated:Copy s3:ObjectCreated:Post s3:ObjectCreated:Put s3:ObjectCreated:PutRetention s3:ObjectCreated:PutLegalHold s3:ObjectCreated:PutTagging s3:Replication:OperationCompletedReplication s3:Replication:OperationFailedReplication]
2021-01-28T12:40:15.6066661Z --- FAIL: TestNameExpand (0.00s)
```

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -1147,6 +1147,13 @@ func (er erasureObjects) addPartial(bucket, object, versionID string) {
 
 // PutObjectTags - replace or add tags to an existing object
 func (er erasureObjects) PutObjectTags(ctx context.Context, bucket, object string, tags string, opts ObjectOptions) (ObjectInfo, error) {
+	// Lock the object before updating tags.
+	lk := er.NewNSLock(bucket, object)
+	if err := lk.GetLock(ctx, globalOperationTimeout); err != nil {
+		return ObjectInfo{}, err
+	}
+	defer lk.Unlock()
+
 	disks := er.getDisks()
 
 	// Read metadata associated with the object from all disks.

--- a/cmd/object-handlers.go
+++ b/cmd/object-handlers.go
@@ -3293,7 +3293,7 @@ func (api objectAPIHandlers) PutObjectTaggingHandler(w http.ResponseWriter, r *h
 	sendEvent(eventArgs{
 		EventName:    event.ObjectCreatedPutTagging,
 		BucketName:   bucket,
-		Object:       ObjectInfo{Bucket: bucket, Name: object, VersionID: objInfo.VersionID, UserTags: tagsStr},
+		Object:       objInfo,
 		ReqParams:    extractReqParams(r),
 		RespElements: extractRespElements(w),
 		UserAgent:    r.UserAgent(),
@@ -3346,11 +3346,13 @@ func (api objectAPIHandlers) DeleteObjectTaggingHandler(w http.ResponseWriter, r
 		opts.UserDefined = make(map[string]string)
 		opts.UserDefined[xhttp.AmzBucketReplicationStatus] = replication.Pending.String()
 	}
+
 	// Delete object tags
 	if err = objAPI.DeleteObjectTags(ctx, bucket, object, opts); err != nil {
 		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
 		return
 	}
+	oi.UserTags = ""
 
 	if opts.VersionID != "" {
 		w.Header()[xhttp.AmzVersionID] = []string{opts.VersionID}
@@ -3365,7 +3367,7 @@ func (api objectAPIHandlers) DeleteObjectTaggingHandler(w http.ResponseWriter, r
 	sendEvent(eventArgs{
 		EventName:    event.ObjectCreatedDeleteTagging,
 		BucketName:   bucket,
-		Object:       ObjectInfo{Bucket: bucket, Name: object, VersionID: oi.VersionID, UserTags: ""},
+		Object:       oi,
 		ReqParams:    extractReqParams(r),
 		RespElements: extractRespElements(w),
 		UserAgent:    r.UserAgent(),

--- a/cmd/object-handlers.go
+++ b/cmd/object-handlers.go
@@ -3265,17 +3265,23 @@ func (api objectAPIHandlers) PutObjectTaggingHandler(w http.ResponseWriter, r *h
 		opts.UserDefined[xhttp.AmzBucketReplicationStatus] = replication.Pending.String()
 	}
 
+	objInfo, err := objAPI.GetObjectInfo(ctx, bucket, object, opts)
+	if err != nil {
+		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+		return
+	}
+
+	tagsStr := tags.String()
+
 	// Put object tags
-	err = objAPI.PutObjectTags(ctx, bucket, object, tags.String(), opts)
+	err = objAPI.PutObjectTags(ctx, bucket, object, tagsStr, opts)
 	if err != nil {
 		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
 		return
 	}
 
 	if replicate {
-		if objInfo, err := objAPI.GetObjectInfo(ctx, bucket, object, opts); err == nil {
-			scheduleReplication(ctx, objInfo, objAPI, sync)
-		}
+		scheduleReplication(ctx, objInfo, objAPI, sync)
 	}
 
 	if opts.VersionID != "" {
@@ -3283,6 +3289,17 @@ func (api objectAPIHandlers) PutObjectTaggingHandler(w http.ResponseWriter, r *h
 	}
 
 	writeSuccessResponseHeadersOnly(w)
+
+	sendEvent(eventArgs{
+		EventName:    event.ObjectCreatedPutTagging,
+		BucketName:   bucket,
+		Object:       ObjectInfo{Bucket: bucket, Name: object, VersionID: objInfo.VersionID, UserTags: tagsStr},
+		ReqParams:    extractReqParams(r),
+		RespElements: extractRespElements(w),
+		UserAgent:    r.UserAgent(),
+		Host:         handlers.GetSourceIP(r),
+	})
+
 }
 
 // DeleteObjectTaggingHandler - DELETE object tagging
@@ -3344,6 +3361,16 @@ func (api objectAPIHandlers) DeleteObjectTaggingHandler(w http.ResponseWriter, r
 	}
 
 	writeSuccessNoContent(w)
+
+	sendEvent(eventArgs{
+		EventName:    event.ObjectCreatedDeleteTagging,
+		BucketName:   bucket,
+		Object:       ObjectInfo{Bucket: bucket, Name: object, VersionID: oi.VersionID, UserTags: ""},
+		ReqParams:    extractReqParams(r),
+		RespElements: extractRespElements(w),
+		UserAgent:    r.UserAgent(),
+		Host:         handlers.GetSourceIP(r),
+	})
 }
 
 // RestoreObjectHandler - POST restore object handler.

--- a/pkg/event/name.go
+++ b/pkg/event/name.go
@@ -41,6 +41,8 @@ const (
 	ObjectCreatedPut
 	ObjectCreatedPutRetention
 	ObjectCreatedPutLegalHold
+	ObjectCreatedPutTagging
+	ObjectCreatedDeleteTagging
 	ObjectRemovedAll
 	ObjectRemovedDelete
 	ObjectRemovedDeleteMarkerCreated
@@ -77,6 +79,7 @@ func (name Name) Expand() []Name {
 			ObjectCreatedCompleteMultipartUpload, ObjectCreatedCopy,
 			ObjectCreatedPost, ObjectCreatedPut,
 			ObjectCreatedPutRetention, ObjectCreatedPutLegalHold,
+			ObjectCreatedPutTagging, ObjectCreatedDeleteTagging,
 			ObjectReplicationComplete, ObjectReplicationFailed,
 		}
 	case ObjectRemovedAll:
@@ -134,6 +137,10 @@ func (name Name) String() string {
 		return "s3:ObjectCreated:Post"
 	case ObjectCreatedPut:
 		return "s3:ObjectCreated:Put"
+	case ObjectCreatedPutTagging:
+		return "s3:ObjectCreated:PutTagging"
+	case ObjectCreatedDeleteTagging:
+		return "s3:ObjectCreated:DeleteTagging"
 	case ObjectCreatedPutRetention:
 		return "s3:ObjectCreated:PutRetention"
 	case ObjectCreatedPutLegalHold:
@@ -244,6 +251,10 @@ func ParseName(s string) (Name, error) {
 		return ObjectCreatedPutRetention, nil
 	case "s3:ObjectCreated:PutLegalHold":
 		return ObjectCreatedPutLegalHold, nil
+	case "s3:ObjectCreated:PutTagging":
+		return ObjectCreatedPutTagging, nil
+	case "s3:ObjectCreated:DeleteTagging":
+		return ObjectCreatedDeleteTagging, nil
 	case "s3:ObjectRemoved:*":
 		return ObjectRemovedAll, nil
 	case "s3:ObjectRemoved:Delete":

--- a/pkg/event/name_test.go
+++ b/pkg/event/name_test.go
@@ -31,8 +31,9 @@ func TestNameExpand(t *testing.T) {
 		{BucketCreated, []Name{BucketCreated}},
 		{BucketRemoved, []Name{BucketRemoved}},
 		{ObjectAccessedAll, []Name{ObjectAccessedGet, ObjectAccessedHead, ObjectAccessedGetRetention, ObjectAccessedGetLegalHold}},
-		{ObjectCreatedAll, []Name{ObjectCreatedCompleteMultipartUpload, ObjectCreatedCopy,
-			ObjectCreatedPost, ObjectCreatedPut, ObjectCreatedPutRetention, ObjectCreatedPutLegalHold, ObjectReplicationComplete, ObjectReplicationFailed}},
+		{ObjectCreatedAll, []Name{ObjectCreatedCompleteMultipartUpload, ObjectCreatedCopy, ObjectCreatedPost, ObjectCreatedPut,
+			ObjectCreatedPutRetention, ObjectCreatedPutLegalHold, ObjectCreatedPutTagging, ObjectCreatedDeleteTagging,
+			ObjectReplicationComplete, ObjectReplicationFailed}},
 		{ObjectRemovedAll, []Name{ObjectRemovedDelete, ObjectRemovedDeleteMarkerCreated}},
 		{ObjectAccessedHead, []Name{ObjectAccessedHead}},
 	}

